course_contents/2_intro_to_python/notes/04. Destructuring.md

```diff
@@ -0,0 +1,49 @@
+### Unpacking
+- unpacking is used to assign values from a collection to variables in a single statement
+- syntax for unpacking:
+
+```py
+x, y, z = collection
+```
+
+- the collection can be a list, tuple, string, etc.
+
+An example of unpacking a list:
+
+```py
+example_list = [1,2,3]
+a, b, c = example_list
+print(a, b, c)
+```
+
+### The `enumerate()` method
+- the `enumerate()` method is used to retrieve an object containing both element and its index inside the collection that is being iterated over
+- syntax for the `enumerate()` method with unpacking:
+
+```py
+for index, element in enumerate(collecion):
+    block of code
+```
+
+An example of the `enumerate()` method:
+
+```py
+collection = ['a', 'b', 'c']
+for index, element in enumerate(collection):
+    print(f"Index:{index}; Element:{element}")
+```
+
+### The `zip()` function
+- the `zip()` function is used to combine two iterables into a single one
+- new iterable is created by joining i-th element of each iterable until the end of the shorter one is reached
+- the `zip()` function returns a `zip` object
+- syntax for the `zip()` function:
```

Comment:
Syntax is missing

Reply:
Nice catch! That must be one of the parts where I worked at late night :), I'll change that

Reply:
I think that lines 39 to 45 aren't _really_ needed, since lines 47-51 already give us the example. The only other thing to mention is that you _can_ zip more than 2 collections.

Maybe those lines could be replaced by a diagram!

Reply:
Added the design note!